### ISSUE TITLE
"multi device" tests get skipped in standard CI

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#92173]
## Concern
The general concern is that people think they're adding tests to CI but these tests are getting skipped, which is bamboozling. Specifically, I wonder how many tests there are across our codebase that require multiple GPUs but are NOT on the multigpu shard, meaning they are always skipped (our runners only have 1 GPU). 

This stemmed from my noticing the `@deviceCountAtLeast(2)` decorator while reviewing a PR, and then a brief search revealed that there are 18 instances of this decorator throughout our codebase. There may other ways people guard/skip tests based on # devices (such as with if statements/with TEST_MULTIGPU). 

To add another layer to this onion, even if people knew about the multigpu shard and how to add a test to it, the multigpu config runs on a periodic cadence, so people may not realize that these tests they added never even ran on the CI of their PR.

## Potential Next Steps
1. Document in deviceCountAtLeast that these do NOT run in most CI, and that tests with this decorator should be added to https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/multigpu-test.sh
2. Run existing unrun tests like `test_clip_grad_norm_multi_device` on the multigpu config
3. I understand that the multigpu shard was historically hardware-constrained. @pytorch/pytorch-dev-infra Is there a near future where multigpu can run on trunk instead of periodic?

**Comment by ngimel** [comment]
cc @wconstab, we do have a multi-gpu build that's run on every PR?

**Comment by janeyx99** [comment]
Wait which?

**Comment by ngimel** [comment]
Hm, maybe I misunderstood Will, and it's just inductor tests that run on a multi-gpu periodic job

**Comment by wconstab** [comment]
its just in the inductor tests, which I added support for here:
https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/test.sh#L796
https://github.com/pytorch/pytorch/blob/master/.jenkins/pytorch/test.sh#L249

you can see it running on PRs if you string search the HUD page for `inductor_distributed`

**Comment by wconstab** [comment]
well, hold on i'm not saying inductor is the only thing that runs multi gpu.  I think there are also PT-D test running multi-gpu, but I added a config that uses newer GPUs and multi-gpu, to make inductor happy.

**Comment by janeyx99** [comment]
Right, but for non-inductor/standard CI, there is only one multigpu config, and most people probably don't realize that they have to specifically put their tests in a distinct script in order for the multi-device tests to run. 

I'm wondering if we now have capacity for more multigpu tests (the current config only run periodically vs on each commit). cc'ing @seemethere

**Comment by wanchaol** [comment]
@wconstab does the inductor GPU tests runs on two GPUs currently? is it using the multigpu shard or using a different approach?

**Comment by wanchaol** [comment]
Yeah I also really feel the pain of this, adding "periodic" label to each PR is something ppl often forget and only after landing a few changes then we notice it breaks some tests for days, which is not nice. Taking https://github.com/pytorch/pytorch/issues/92343 as a example, this is the second time I landed a PR without noticing it breaks the cuda tests with multigpus.

In case of hardware constraints, I'm wondering if we can at least have some non-fancy GPUs (i.e. 4 or 8) that are cheaper and we could enable tests on it for every PR?

**Comment by seemethere** [comment]
> In case of hardware constraints, I'm wondering if we can at least have some non-fancy GPUs (i.e. 4 or 8) that are cheaper and we could enable tests on it for every PR?

The problem here isn't cost, the problem is _capacity_. If we enable a GPU workflow that requires an instance with 4-8 GPUs those instances are just _less available_ in AWS meaning we need to run an extremely limited set of tests on those (<30 minutes) or need to accept that those jobs will queue for a long time (which is probably not acceptable).

We're not afraid of throwing money at the problem if it's worth it but multi-gpu tests on every PR seems like a bit of an unnecessary step to me.

**Comment by wanchaol** [comment]
> The problem here isn't cost, the problem is _capacity_. If we enable a GPU workflow that requires an instance with 4-8 GPUs those instances are just _less available_ in AWS meaning we need to run an extremely limited set of tests on those (<30 minutes) or need to accept that those jobs will queue for a long time (which is probably not acceptable).
> 
> We're not afraid of throwing money at the problem if it's worth it but multi-gpu tests on every PR seems like a bit of an over-correction to me.

I see, make sense, thanks for explaining @seemethere! I think I'm fine staying with the current approach, I'm wondering if we could auto label "periodic" for changes to certain files, so when we changing certain folders it would trigger multi-gpu tests? i.e. some of the distributed tests would like to run a multi-gpu tests on the PR that changes a certain folder, would it make sense for us to do some sort of auto labelling for these cases?

**Comment by seemethere** [comment]
> I'm wondering if we could auto label "periodic" for changes to certain files, so when we changing certain folders it would trigger multi-gpu tests? i.e. some of the distributed tests would like to run a multi-gpu tests on the PR that changes a certain folder, would it make sense for us to do some sort of auto labelling for these cases?

@huydhn @clee2000 do we have a mechanism to enable this with ciflow? We could possibly even put it multi-gpu tests into their own workflow similar to what we do with inductor?

**Comment by ZainRizvi** [comment]
If we have specific files/folders that should always trigger periodic builds, we could use auto-labeler to tag those PRs with the periodic label and kick off those builds
https://github.com/pytorch/pytorch/blob/master/.github/labeler.yml
